library/core/src/alloc/mod.rs

```diff
@@ -114,6 +116,10 @@ pub unsafe trait Allocator {
     /// The returned block may have a larger size than specified by `layout.size()`, and may or may
     /// not have its contents initialized.
     ///
+    /// The returned block of memory remains valid as long as it is [*currently allocated*] and the shorter of:
+    ///   - the borrow-checker lifetime of the allocator type itself.
```

Comment:
I wonder if we should phrase or give a heuristic that this is equivalent to treating the allocation as borrowing from the `&self` argument - i.e., the returned NonNull is really closer to a `&'a mut [MaybeUninit<u8>]`, in some sense.

But this seems OK for now.